### DISCUSSION TITLE
Mocking setter from useAtom in jest

### DISCUSSION THREAD

**Discussion by pmilaszewski** [pmndrs/jotai#1926] [Q&A]
Hi, I have a question about mocking setter from useAtom hook. I'm using @testing-library/react-native for testing my project. I my component I have something like: 
`const [data, setData] = useAtom(dataAtom)`
and I have a button which should call this `setData`
`<Button onPress={setData} testID={'button'}>Set new data</Button>`
so in my .test file I want to check if this method have been called after clicking this button.

Is it possible to mock this `setData` in any way?

**Comment by dai-shi** [comment]
Hi, how would you if it's `useState`? If there's best practice, we could follow it.
Otherwise, mocking dataAtom entirely might be an option.

**Reply by pmilaszewski** [reply]
@dai-shi Could you please provide any examples? Unfortunately I tried many ways and it didn't work. I also tried to implement it as here https://stackoverflow.com/a/65334852 but it also failed. I'll be glad for help or explanation if it's not possible to test.

**Reply by pmilaszewski** [reply]
Tried also something like that
`it('Should call setCart after press button', async () => {

    const { getByTestId } = render(<TestComponent />)
    const { result } = renderHook(() => useAtom(cartAtom))
    const [_, mockedSetCart] = result.current
    const button = getByTestId('Cart_button')
    fireEvent.press(button)
    await waitFor(() => {
      expect(mockedSetCart).toHaveBeenCalledWith([])
    })
  })
`
  but I received an error:
   Matcher error: received value must be a mock or spy function
    Received has type:  function
    Received has value: [Function anonymous]

**Reply by dai-shi** [reply]
> how would you if it's `useState`? 

I asked it because I don't have an answer.
If what you are trying is impossible with `useState`, it can't be done with `useAtom`, I suspect.

**Reply by arjunvegda** [reply]
@pmilaszewski  - Another approach is to test the outcome. Rather than checking if `setData` was called, we could check if UI changed with the expected data. 😃 

Kent has a great write-up on this - https://kentcdodds.com/blog/how-to-know-what-to-test#remembering-why-we-test
